DataSetExplorer/DataSetBuilder/FullDataSetBuilderService.cs

```diff
@@ -0,0 +1,70 @@
+﻿using CodeModel;
+using CodeModel.CaDETModel;
+using DataSetExplorer.DataSetBuilder.Model;
+using DataSetExplorer.DataSetSerializer;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.Linq;
+
+namespace DataSetExplorer
+{
+    class FullDataSetBuilderService : IFullDataSetBuilder
+    {
+        public IEnumerable<IGrouping<string, DataSetInstance>> GetAnnotatedInstancesGroupedBySmells(int? annotatorId)
+        {
+            var allAnnotatedInstances = new List<DataSetInstance>();
+
+            var projects = GetAnnotatedProjects();
+            foreach (var key in projects.Keys)
+            {
+                CodeModelFactory factory = new CodeModelFactory();
+                CaDETProject project = factory.CreateProjectWithCodeFileLinks(key.ToString());
+
+                var importer = new ExcelImporter(projects[key].ToString());
+                var annotatedInstances = importer.Import("Clean CaDET").GetAllInstances();
+
+                LoadAnnotators(ref annotatedInstances);
+                if (annotatorId != null) annotatedInstances = annotatedInstances.Where(i => i.IsAnnotatedBy((int)annotatorId)).ToList();
+                allAnnotatedInstances.AddRange(FillInstancesWithMetrics(annotatedInstances, project));
+            }
+            return allAnnotatedInstances.GroupBy(i => i.Annotations.ToList()[0].InstanceSmell.Value);
+        }
+
+        private List<DataSetInstance> FillInstancesWithMetrics(List<DataSetInstance> annotatedInstances, CaDETProject project)
+        {
+            return annotatedInstances.Select(i => {
+                i.MetricFeatures = project.GetMetricsForCodeSnippet(i.CodeSnippetId);
+                return i;
+            }).ToList();
+        }
+
+        private ListDictionary GetAnnotatedProjects()
```

Comment:
Why not a list or set?

Reply:
I used a ListDictionary to save two pieces of information for each project (e.g., path to the local repository and path to the folder with annotations).